framework/web/Request.php

```diff
@@ -1131,15 +1142,55 @@ public function getUserAgent()
      */
     public function getUserIP()
     {
-        foreach ($this->ipHeaders as $ipHeader) {
+        foreach($this->getTrustedIpHeaders() as $ipHeader) {
             if ($this->headers->has($ipHeader)) {
-                return trim(explode(',', $this->headers->get($ipHeader))[0]);
+                $ip = $this->getUserIpFromIpHeader($this->headers->get($ipHeader));
+                if(isset($ip)) {
+                    return $ip;
+                }
             }
         }
 
         return $this->getRemoteIP();
     }
 
+    /**
+     * Return user IP's from IP header.
+     * @param string $ips comma separated IP list
+     * @return string|null
+     * @see $getUserHost
+     * @see $ipHeader
+     * @see $trustedHeaders
+     * @since 2.0.28
+     */
+    protected function getUserIpFromIpHeader($ips)
+    {
+        $ips = preg_split('/\s*,\s*/', trim($ips));
+        krsort($ips);
+        $validator = $this->getIpValidator();
+        $resultIp = null;
+        $stop = false;
+        foreach ($ips as $ip) {
+            foreach ($this->trustedHosts as $cidr => $cidrOrHeaders) {
+                if (!is_array($cidrOrHeaders)) {
+                    $cidr = $cidrOrHeaders;
+                }
+                $validator->setRanges($cidr);
+                $resultIp = $ip;
+                if ($validator->validate($ip)) {
```

Comment:
I think we don't need `$stop` at all:

```php
if (!$validator->validate($ip)) {
    break 2;
}
```

Reply:
`$stop` is required because it is within the cycle that it will be overwritten. If it find a IP address in a subsequent iteration of the loop, `$stop=false` and the internal loop will end.

I simplified the code.

Reply:
Maybe renaming this variable (`$isValid`?) will help - right now it is little hard to understand what is happening here, because `$stop` does not really explaining why it stops.

Reply:
variable is renamed